src/H5FDmpio.c

```diff
@@ -1745,6 +1747,61 @@ H5FD__mpio_truncate(H5FD_t *_file, hid_t H5_ATTR_UNUSED dxpl_id, hbool_t H5_ATTR
     FUNC_LEAVE_NOAPI(ret_value)
 } /* end H5FD__mpio_truncate() */
 
+/*-------------------------------------------------------------------------
+ * Function:    H5FD__mpio_delete
+ *
+ * Purpose:     Delete a file
+ *
+ * Return:      SUCCEED/FAIL
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5FD__mpio_delete(const char *filename, hid_t fapl_id)
+{
+    H5P_genplist_t *plist; /* Property list pointer */
+    MPI_Comm        comm     = MPI_COMM_NULL;
+    MPI_Info        info     = MPI_INFO_NULL;
+    int             mpi_rank = INT_MAX;
+    int             mpi_code;            /* MPI return code */
+    herr_t          ret_value = SUCCEED; /* Return value */
+
+    FUNC_ENTER_STATIC
+
+    HDassert(filename);
+
+    if (NULL == (plist = H5P_object_verify(fapl_id, H5P_FILE_ACCESS)))
+        HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a file access property list")
+
+    /* Get the MPI communicator and info from the fapl */
+    if (H5P_FILE_ACCESS_DEFAULT != fapl_id && H5FD_MPIO == H5P_peek_driver(plist)) {
```

Comment:
This code is "fence-sitting" - Either we “know” that the FAPL is non-default (because we arrived in the callback) or we need to protect the callback for the situation where the FAPL doesn’t have the MPI-IO VFD as the value from peek_driver.  Right now, the code looks like you could get to the callback with the default FAPL and then (if that theoretically occurred) you could call MPI_Comm_rank with MPI_COMM_NULL and have a very weird value for the mpi_rank variable.  I would just take out the check for a non-default FAPL and assert that you have the H5FD_MPIO driver, then you would be “safe” to assume a valid MPI communicator and a valid mpi_rank.

Reply:
I'll make this change.